src/Workspaces/Core/Portable/Rename/Renamer.RenameDocumentActionSet.cs

```diff
@@ -0,0 +1,120 @@
+
+            internal RenameDocumentActionSet(
+                ImmutableArray<RenameDocumentAction> actions,
+                ProjectId projectId,
+                DocumentId documentId,
+                string documentName,
+                IReadOnlyList<string> documentFolders)
+            {
+                ApplicableActions = actions;
+                _documentFolders = documentFolders;
+                _documentId = documentId;
+                _documentName = documentName;
+                _projectId = projectId;
+            }
+
+            /// <summary>
+            /// All applicable actions computed for the action
+            /// </summary>
+            public ImmutableArray<RenameDocumentAction> ApplicableActions { get; }
+
+            /// <summary>
+            /// Same as calling <see cref="UpdateSolutionAsync(Solution, ImmutableArray{RenameDocumentAction}, CancellationToken)"/> with 
+            /// <see cref="ApplicableActions"/> as the argument
+            /// </summary>
+            public Task<Solution> UpdateSolutionAsync(Solution solution, CancellationToken cancellationToken)
+                => UpdateSolutionAsync(solution, ApplicableActions, cancellationToken);
+
+            /// <summary>
+            /// Applies each <see cref="RenameDocumentAction"/> in order and returns the final solution. 
+            /// All actions must be contained in <see cref="ApplicableActions" />
+            /// </summary>
+            /// <remarks>
+            /// An empty action set is still allowed and will return a modified solution
+            /// that will update the document properties as appropriate.
+            /// </remarks>
+            public async Task<Solution> UpdateSolutionAsync(Solution solution, ImmutableArray<RenameDocumentAction> actions, CancellationToken cancellationToken)
+            {
+                if (actions.Any(a => !ApplicableActions.Contains(a)))
+                {
+                    throw new InvalidOperationException("Cannot apply action that is not in ApplicableActions");
+                }
+
+                // Prior to updating the solution it's possible the document id has changed between the time 
+                // the document action info was generated and when the solution update is applied. We
+                // do a best effort to still locate the document if the id has changed.
+                var document = GetDocument(solution);
+
+                // If the document was found in the solution then the current id will be durable across actions
+                // since we own the solution snapshot at this point. 
+                var documentId = document.Id;
+
+                // Make sure that the document name and folders are updated to what we expect them to be
+                solution = solution
+                    .WithDocumentName(documentId, _documentName)
+                    .WithDocumentFolders(documentId, _documentFolders);
```

Comment:
so these apply to *all* actoins?

Reply:
Yes. Just because there is no _extra_ work to be done, I didn't want the API to be wrong. 

Example: 

```csharp
var renameInfo = await Renamer.RenameDocumentName(document, newname, ...);
var solution = renameInfo.UpdateSolutionAsync(solution, cancellationToken); 
```

This code should always update the document name, even if there's no symbol changes to be made (or header changes in the future)